jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -486,60 +239,193 @@ else if (!Files.isDirectory(path) && !FileID.isWebArchive(path))
             }
 
             //set a backup value for the path to the war in case it hasn't already been set
-            properties.put(Deployable.WAR, path.toString());
-            return initializeContextHandler(context, path, properties);
+            appAttributes.setAttribute(Deployable.WAR, path.toString());
+            return initializeContextHandler(context, path, appAttributes);
         }
         finally
         {
             Thread.currentThread().setContextClassLoader(old);
         }
     }
 
-    protected Object applyXml(Object context, Path xml, String environment, Map<String, String> properties) throws Exception
+    /**
+     * Get the default {@link Environment} name for discovered web applications that
+     * do not declare the {@link Environment} that they belong to.
+     *
+     * <p>
+     * Falls back to {@link Environment#getAll()} list, and returns
+     * the first name returned after sorting with {@link Deployable#ENVIRONMENT_COMPARATOR}
+     * </p>
+     *
+     * @return the default environment name.
+     */
+    public String getDefaultEnvironmentName()
+    {
+        if (defaultEnvironmentName == null)
+        {
+            return Environment.getAll().stream()
+                .map(Environment::getName)
+                .max(Deployable.ENVIRONMENT_COMPARATOR)
+                .orElse(null);
+        }
+        return defaultEnvironmentName;
+    }
+
+    public void setDefaultEnvironmentName(String name)
+    {
+        this.defaultEnvironmentName = name;
+    }
+
+    @Deprecated
+    public Map<String, String> getProperties(Environment environment)
+    {
+        return asProperties(environment);
+    }
+
+    public void loadProperties(Environment environment, Path path) throws IOException
+    {
+        try (InputStream inputStream = Files.newInputStream(path))
+        {
+            loadProperties(environment, inputStream);
+        }
+    }
+
+    public void loadProperties(Environment environment, Resource resource) throws IOException
+    {
+        try (InputStream inputStream = IOResources.asInputStream(resource))
+        {
+            loadProperties(environment, inputStream);
+        }
+    }
+
+    public void loadPropertiesFromString(Environment environment, String path) throws IOException
+    {
+        loadProperties(environment, Path.of(path));
+    }
+
+    /**
+     * Configure the Environment specific Deploy settings.
+     *
+     * @param name the name of the environment.
+     * @return the deployment configuration for the {@link Environment}.
+     */
+    public EnvironmentConfig configureEnvironment(String name)
```

Comment:
This is strange, as calling this method returns a different object every time, so calling it twice make unclear whether the previous configuration was retained or not.

Since `EnvironmentConfig` is just a wrapper around `Environment` properties (to give compile-time checks), it should be stored as a property itself in the `Environment`.

This method should be renamed `getDeploymentConfiguration(String name)`, and it is not really an `EnvironmentConfig`, it is the `DeploymentConfiguration` for that specific environment.